lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1667,41 +1667,41 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
     bool IsConst = BVar->isConstant();
     llvm::GlobalValue::LinkageTypes LinkageTy = transLinkageType(BVar);
     SPIRVStorageClassKind BS = BVar->getStorageClass();
-    Constant *Initializer = nullptr;
     SPIRVValue *Init = BVar->getInitializer();
-    if (Init)
-      Initializer = dyn_cast<Constant>(transValue(Init, F, BB, false));
-    else if (LinkageTy == GlobalValue::CommonLinkage)
-      // In LLVM, variables with common linkage type must be initialized to 0.
-      Initializer = Constant::getNullValue(Ty);
-    else if (BS == SPIRVStorageClassKind::StorageClassWorkgroup)
-      Initializer = dyn_cast<Constant>(UndefValue::get(Ty));
-    else if ((LinkageTy != GlobalValue::ExternalLinkage) &&
-             (BS == SPIRVStorageClassKind::StorageClassCrossWorkgroup))
-      Initializer = Constant::getNullValue(Ty);
-
     if (BS == StorageClassFunction && !Init) {
       assert(BB && "Invalid BB");
       return mapValue(BV, new AllocaInst(Ty, 0, BV->getName(), BB));
     }
-    SPIRAddressSpace AddrSpace;
 
+    SPIRAddressSpace AddrSpace;
     bool IsVectorCompute =
         BVar->hasDecorate(DecorationVectorComputeVariableINTEL);
+    Constant *Initializer = nullptr;
     if (IsVectorCompute) {
       AddrSpace = VectorComputeUtil::getVCGlobalVarAddressSpace(BS);
-      if (!Initializer)
-        Initializer = UndefValue::get(Ty);
+      Initializer = UndefValue::get(Ty);
     } else
       AddrSpace = SPIRSPIRVAddrSpaceMap::rmap(BS);
-
-    auto LVar = new GlobalVariable(*M, Ty, IsConst, LinkageTy, Initializer,
-                                   BV->getName(), 0,
+    auto LVar = new GlobalVariable(*M, Ty, IsConst, LinkageTy,
+                                   /*Initializer=*/nullptr, BV->getName(), 0,
                                    GlobalVariable::NotThreadLocal, AddrSpace);
+    auto Res = mapValue(BV, LVar);
+    if (Init)
+      Initializer = dyn_cast<Constant>(transValue(Init, F, BB, false));
+    else if (LinkageTy == GlobalValue::CommonLinkage)
+      // In LLVM, variables with common linkage type must be initialized to 0.
+      Initializer = Constant::getNullValue(Ty);
+    else if (BS == SPIRVStorageClassKind::StorageClassWorkgroup)
+      Initializer = dyn_cast<Constant>(UndefValue::get(Ty));
+    else if ((LinkageTy != GlobalValue::ExternalLinkage) &&
+             (BS == SPIRVStorageClassKind::StorageClassCrossWorkgroup))
```

Comment:
That would look nasty, but I guess we also need to check for StorageClassDeviceOnlyINTEL and StorageClassHostOnlyINTEL added in https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/579/files

Reply:
The only thing I did here is move. So, they weren't checked before too. I think it is a change that we need to make separately.

Reply:
Agree